### ISSUE TITLE
Windows path definitions and shared libraries

### ISSUE THREAD

**Issue by nyorain** [mesonbuild/meson#2105]
Hey, first of all thank you for this whole project, was a pleasure to use to far switching my projects from cmake to meson! :)
But there are some struggles i am having with getting projects to work on windows.
I am using the latest meson release (0.41.1)

# Passing paths as definition

It might sometimes be needed to pass a path as definition to a compiler.
On unix this would be done using something like
```
add_project_arguments(
	'-DSOME_PATH="@0@"'.format(meson.source_root()),
	language: 'cpp')
```
This does not work on windows since meson.source_root() returns a path with backslashes on windows.
I have no idea how to get this to work. Would probably require something like an 'escape' string function (that transforms '\' into '\\'). Using cmake this works, it seems to convert/escape the path somehow (and if it would not, one could hack around it using its regex string replace operations which is not possible using meson).

# Shared libraries in build path

This issue is probably known, but i have no idea as well on how to handle this. Would be nice to document workarounds on the meson homepage.

When building a shared library on windows and then an executable linking to it in a different dir, the executable will not find the shared library when executed (when executed in build path, not installed).
I know this is because of windows providing no rpath mechanism but then it should be possible at least to output libraries and executables on windows (or generally) into the same build directory (the usual way to work around this using cmake, CMAKE_RUNTIME_OUTPUT_DIRECTORY).

I feel like this is a rather huge problem at the moment (if windows is a priority at all), there is basically no way to get a project building a shared library and then tests or examples for it to work except by copying the library everytime to the example/test directory manually (meson could also automate this).
I don't think installing the project everytime after building and then run from the builddir is a solution.

# Vulkan

Using vulkan as dependency is somewhat tricky on windows (since we cannot simply access vulkan sdk environment variable). I read about it in another thread, would something like vulkan dependency module built into meson still be appreciated? Could try to figure it out if so.

**Comment by jpakkane** [comment]
> Using cmake this works, it seems to convert/escape the path somehow

We should do that too but apparently not. That is a bug. However it is recommended that you try to minimize the number of compiler arguments and instead create a configuration file and write your system data there. This reduces clutter and makes things easier to inspect and also for IDEs.

> I know this is because of windows providing no rpath mechanism

Turns out there is a way to fake this with executable manifest files. There has been talk of adding support for those. In the mean time you can set the `layout` option to value `flat` which puts all output in one directory.

> would something like vulkan dependency module built into meson still be appreciated?

If using vulkan can't be done in an easy way otherwise, sure.

**Comment by nyorain** [comment]
> However it is recommended that you try to minimize the number of compiler arguments and instead create a configuration file and write your system data there

Yeah, using a configuration file is a good idea but it seems to have the same bug on windows.
Still something like ```#define SOME_PATH "C:\some_path"``` which results in a compile time error.

Didn't know about --layout=flat, thank you. Would probably a good idea to add this to the manual or 'How do i do X in meson' on the homepage i guess (with a note that it might be needed on windows atm). 
Seems to be what i needed but while using it I stumbled about #1993, a different issue though.

Will start working myself into the meson codebase for the vulkan module as soon as i find some time.
Manifests as rpath substitute look promising (and ugly, but hey it's windows...) as well.

**Comment by jpakkane** [comment]
Configuration has the `set_quoted` convenience function that should do the quoting. If it doesn't then that's a bug.

Regardless, having `\` in compilation arguments should work. [we even have a test for it](https://github.com/mesonbuild/meson/tree/master/test%20cases/common/115%20spaces%20backslash).

**Comment by nyorain** [comment]
I used set_quoted. The problem is when passing a path built inside meson (NOT hardcoded into the meson.build file) to the compiler (definition or coniguration file). The test escapes backslashes manually that is why it works. But when using a path built by meson (e.g. join_paths or meson.source_root()) there is no way to replace ```\``` with ```\\```.
And the c compiler thinks it is an invalid escape code.

**Comment by QuLogic** [comment]
@jpakkane `set_quoted` only escapes double quotes; it should probably do a bit more.
